[CT-3514] [Feature] Use formatting for `package-lock.yml` that passes `yamlint` linting checks

### Is this a new bug in dbt-core?

- [X] I believe this is a new bug in dbt-core
- [X] I have searched the existing issues, and I could not find an existing issue for this bug

### Current Behavior

`package-lock.yml` does not have a document separator and does not have correct indentation.

Run `ymllint --strict` on a newly created `package-lock.yml`.

```bash
> dbt deps
> cat package-lock.yml
packages:
- package: dbt-labs/audit_helper
  version: 0.9.0
- package: dbt-labs/codegen
  version: 0.12.1
- package: dbt-labs/dbt_utils
  version: 1.1.1
sha1_hash: 3cf6af6c0a4ee17cf53205ae383d515915b55f5f
> docker run --rm -v $(pwd):/data cytopia/yamllint:latest --strict /data/package-lock.yml
/data/package-lock.yml
  1:1       warning  missing document start "---"  (document-start)
  2:1       error    wrong indentation: expected at least 1  (indentation)

> echo $?
1
```

### Expected Behavior

That it be generated correctly formatted. For example here, when manually fixed it passes.

```bash
> cat package-lock.yml
---

packages:
  - package: dbt-labs/audit_helper
    version: 0.9.0
  - package: dbt-labs/codegen
    version: 0.12.1
  - package: dbt-labs/dbt_utils
    version: 1.1.1
sha1_hash: 3cf6af6c0a4ee17cf53205ae383d515915b55f5f
> docker run --rm -v $(pwd):/data cytopia/yamllint:latest --strict /data/package-lock.yml
> echo $?
0
```

### Steps To Reproduce

As above.

### Relevant log output

```shell
N/A
```


### Environment

```markdown
- OS: Linux
- Python: 3.11
- dbt: 1.7.4
```


### Which database adapter are you using with dbt?

snowflake

### Additional Context

_No response_

--- a/.changes/unreleased/Fixes-20240106-003649.yaml
+++ b/.changes/unreleased/Fixes-20240106-003649.yaml
@@ -1,0 +1,6 @@
+kind: Fixes
+body: fix lock-file bad indentation
+time: 2024-01-06T00:36:49.547533+09:00
+custom:
+  Author: jx2lee
+  Issue: "9319"

--- a/core/dbt/task/deps.py
+++ b/core/dbt/task/deps.py
@@ -201,7 +201,7 @@ class DepsTask(BaseTask):
         )
 
         with open(lock_filepath, "w") as lock_obj:
-            yaml.safe_dump(packages_installed, lock_obj)
+            yaml.dump(packages_installed, lock_obj, Dumper=dbtPackageDumper)
 
         fire_event(DepsLockUpdating(lock_filepath=lock_filepath))
 

--- a/tests/functional/dependencies/test_dependency_options.py
+++ b/tests/functional/dependencies/test_dependency_options.py
@@ -35,10 +35,10 @@ class TestDepsOptions(object):
         assert (
             contents
             == """packages:
-- package: fivetran/fivetran_utils
-  version: 0.4.7
-- package: dbt-labs/dbt_utils
-  version: 1.1.1
+  - package: fivetran/fivetran_utils
+    version: 0.4.7
+  - package: dbt-labs/dbt_utils
+    version: 1.1.1
 sha1_hash: 71304bca2138cf8004070b3573a1e17183c0c1a8
 """
         )
@@ -52,10 +52,10 @@ sha1_hash: 71304bca2138cf8004070b3573a1e17183c0c1a8
         assert (
             contents
             == """packages:
-- package: fivetran/fivetran_utils
-  version: 0.4.7
-- package: dbt-labs/dbt_utils
-  version: 1.1.1
+  - package: fivetran/fivetran_utils
+    version: 0.4.7
+  - package: dbt-labs/dbt_utils
+    version: 1.1.1
 sha1_hash: 71304bca2138cf8004070b3573a1e17183c0c1a8
 """
         )

--- a/tests/functional/dependencies/test_simple_dependency.py
+++ b/tests/functional/dependencies/test_simple_dependency.py
@@ -241,12 +241,12 @@ class TestSimpleDependencyWithSubdirs(object):
         run_dbt(["deps"])
         assert os.path.exists("package-lock.yml")
         expected = """packages:
-- git: https://github.com/dbt-labs/dbt-multipe-packages.git
-  revision: 53782f3ede8fdf307ee1d8e418aa65733a4b72fa
-  subdirectory: dbt-utils-main
-- git: https://github.com/dbt-labs/dbt-multipe-packages.git
-  revision: 53782f3ede8fdf307ee1d8e418aa65733a4b72fa
-  subdirectory: dbt-date-main
+  - git: https://github.com/dbt-labs/dbt-multipe-packages.git
+    revision: 53782f3ede8fdf307ee1d8e418aa65733a4b72fa
+    subdirectory: dbt-utils-main
+  - git: https://github.com/dbt-labs/dbt-multipe-packages.git
+    revision: 53782f3ede8fdf307ee1d8e418aa65733a4b72fa
+    subdirectory: dbt-date-main
 sha1_hash: b9c8042f29446c55a33f9f211737f445a640c7a1
 """
         with open("package-lock.yml") as fp:

